includes/libs/CSSMin.php

```diff
@@ -50,176 +54,474 @@ class CSSMin {
 				}
+
+				$files[] = $path . $url;
 			}
 		}
 		return $files;
 	}
 
 	/**
-	 * @param $file string
+	 * Encode an image file as a data URI.
+	 *
+	 * If the image file has a suitable MIME type and size, encode it as a data URI, base64-encoded
+	 * for binary files or just percent-encoded otherwise. Return false if the image type is
+	 * unfamiliar or file exceeds the size limit.
+	 *
+	 * @param string $file Image file to encode.
+	 * @param string|null $type File's MIME type or null. If null, CSSMin will
+	 *     try to autodetect the type.
+	 * @param bool $ie8Compat By default, a data URI will only be produced if it can be made short
+	 *     enough to fit in Internet Explorer 8 (and earlier) URI length limit (32,768 bytes). Pass
+	 *     `false` to remove this limitation.
+	 * @return string|bool Image contents encoded as a data URI or false.
+	 */
+	public static function encodeImageAsDataURI( $file, $type = null, $ie8Compat = true ) {
+		// Fast-fail for files that definitely exceed the maximum data URI length
+		if ( $ie8Compat && filesize( $file ) >= self::DATA_URI_SIZE_LIMIT ) {
+			return false;
+		}
+
+		if ( $type === null ) {
+			$type = self::getMimeType( $file );
+		}
+		if ( !$type ) {
+			return false;
+		}
+
+		return self::encodeStringAsDataURI( file_get_contents( $file ), $type, $ie8Compat );
+	}
+
+	/**
+	 * Encode file contents as a data URI with chosen MIME type.
+	 *
+	 * The URI will be base64-encoded for binary files or just percent-encoded otherwise.
+	 *
+	 * @since 1.25
+	 *
+	 * @param string $contents File contents to encode.
+	 * @param string $type File's MIME type.
+	 * @param bool $ie8Compat See encodeImageAsDataURI().
+	 * @return string|bool Image contents encoded as a data URI or false.
+	 */
+	public static function encodeStringAsDataURI( $contents, $type, $ie8Compat = true ) {
+		// Try #1: Non-encoded data URI
+		// The regular expression matches ASCII whitespace and printable characters.
+		if ( preg_match( '/^[\r\n\t\x20-\x7e]+$/', $contents ) ) {
+			// Do not base64-encode non-binary files (sane SVGs).
+			// (This often produces longer URLs, but they compress better, yielding a net smaller size.)
+			$uri = 'data:' . $type . ',' . rawurlencode( $contents );
+			if ( !$ie8Compat || strlen( $uri ) < self::DATA_URI_SIZE_LIMIT ) {
+				return $uri;
+			}
+		}
+
+		// Try #2: Encoded data URI
+		$uri = 'data:' . $type . ';base64,' . base64_encode( $contents );
+		if ( !$ie8Compat || strlen( $uri ) < self::DATA_URI_SIZE_LIMIT ) {
+			return $uri;
+		}
+
+		// A data URI couldn't be produced
+		return false;
+	}
+
+	/**
+	 * Serialize a string (escape and quote) for use as a CSS string value.
+	 * http://www.w3.org/TR/2013/WD-cssom-20131205/#serialize-a-string
+	 *
+	 * @param string $value
+	 * @return string
+	 * @throws Exception
+	 */
+	public static function serializeStringValue( $value ) {
+		if ( strstr( $value, "\0" ) ) {
+			throw new Exception( "Invalid character in CSS string" );
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS112)